### DISCUSSION TITLE
admin role metadata from Auth0 - custom metadata required for rolesField?

### DISCUSSION THREAD

**Discussion by eastendtom** [lowdefy/lowdefy#666] [Questions and Help] [answered]
Hi there team Lowdefy,

Loving this platform. It's a super quick route in to building a database app for a coding noob like me! Thanks so much for this really cool project.

I've hit a stumbling block trying to get users working properly with the repo lowdefy/lowdefy-example-openid-connect

I'm successfully logging into my App with a user via Auth0, and this user has been setup with a role called 'admin' in the Auth0 GUI. However when I log in on openid app, the 'secret' pages which admin should see are still unavailable. Something isn't right, and also I can't edit the Profile with a 1st and 2nd name, not sure if that is related.

Here is what I have in lowdefy.yaml, and I've tried various options but maybe the rolesField needs a string of some sort to resolve?

Thanks in advance,
Tom.

config:
  homePageId: login
  auth:
    openId:
      rolesField: https://pulls.eu.auth0.com/roles/
      logoutRedirectUri: '{{ openid_domain }}/v2/logout?returnTo={{ host }}/logged-out&client_id={{ client_id }}'
    pages:

**Comment by SamTolmay** [comment, marked as answer]
First make sure you have added a login rule, you can find an example here: https://github.com/lowdefy/lowdefy-example-openid-connect/blob/main/auth0/loginRule.js. You will need to configure the MongoDB URI as a secret, and set the namespace.

Another issue  is that your roles field uses the auth0.com domain. You can read more about namespaced claims at https://auth0.com/docs/tokens/create-namespaced-custom-claims, but the auth0 domain is not allowed. You can use any other domain name, it doesn't even need to be a valid domain, so 'http://cajbcnseoidma.com/roles' should be fine (This should be the same as set in the login rule). You also don't want the trailing slash in your rolesField definition.

The login rule reads a role from the user object in the database, so the user document should have a field called roles, that has an array of strings that are the user roles.

To check that your user object has the roles field, you can run `lowdefy.user` in the browser console (The `lowdefy` object is available in the development server, and allows you to inspect the state of your app)

**Reply by eastendtom** [reply]
Thanks Sam! I REALLY appreciate the help.

1. loginRule.js - I have added a namespace as suggested, but I can't see anywhere to input MongoDB URI in loginRule.js, I've configured this in lowdefy.yaml, is it the same code snippet somewhere here? Sorry I'll need more guidance / an example please....

   // The MongoDB connection is memoized in global, so it can be reused.
    // If it does not exist, it is created and stored.
    if (!global.userCollection) {
      const client = new MongoClient(configuration.BWE_MDB_URI, {
        useNewUrlParser: true,
        useUnifiedTopology: true,
      });

2. back to basics...! (sorry for another noob question) how do I run / get to the browser console? If I can read the user object strings coming back from Auth0 then I can hopefully change the custom metadata on the Auth0 side

Thanks so much,
Tom.

**Reply by SamTolmay** [reply]
1. When you create the rule, there is a settings section beneath it. So there you can add MONGODB_URI as the key, and use the same connection you are using for Lowdefy as the value. 
You also want the rule change this to (This is just a typo in the example):
```
const client = new MongoClient(configuration.MONGODB_URI, {
useNewUrlParser: true,
useUnifiedTopology: true,
});
```

2. No worries. I always right-click and chose "inspect", and then select the console tab. There are shortcut keys as well, for Chrome its Shift + CTRL + J or Option + ⌘ + J

**Reply by eastendtom** [reply]
Thanks Sam, that's really helpful. I've updated loginRule.js as you suggest, and managed to run lowdefy.user - thanks for the noob tip!

The output doesn't have role info coming through, so I guess that I need to configure Auth0 to inject this metadata somehow. The web is short on helpful info here. I've tried adding a short line of code to...

Auth0 / User Management / Users / Details / Metadata / User Metadata

{
"role": "admin"
}

But no dice when I log out and log back in, lowdefy.user returns the same lack of 'role' field. Any ideas of how / where to do this in Auth0 GUI, please? 

Cheers!!
Tom.

**Reply by SamTolmay** [reply]
The login rule reads from the MongoDB database, not the Auth0 roles. So you need to add the role on the user object in MongoDB. The new-user and edit-user pages also allow you to set those roles, so if you can't access them, you can disable the authorization on those pages, and give yourself the role. The roles object in MongoDB should be an array of strings, so:
```
{
  "roles": ["admin"]
}
```

**Reply by eastendtom** [reply]
Thank you Sam, that works to disable and get to Lowdefy user pages that way. However when I re-enable authorization the relevent pages are hidden from my Admin user again, despite that user being given the 'admin' role in Lowdefy. So I guess I still need to have Auth0 issue the role metadata correctly?

Would you please be able to post what the console should show if Auth0 role is all correct? I get. the below....

email: "x-x-x@google.com"
email_verified: true
name: "x-x-x@google.com"
nickname: "smiley"
picture: "https://s.g..............
sub: "auth0|60dxxxxxxxxxxxxxxxxx"
updated_at: "2021-06-25T11:51:35.522Z"

**Reply by SamTolmay** [reply]
You would need to have `"http://your_domain.com/roles": ["admin"]` as an additional field on the user object. So if you don't have it, that means the Auth0 login rule isn't adding the role.

**Reply by eastendtom** [reply]
Cheers Sam, that needs to be an additional field on the user object in Auth0, just to be clear? 

And the http://your_domain.com/role URL matches the namespace entries in loginRule.js ?

**Reply by SamTolmay** [reply]
You don't need any fields on the user object in Auth0. The login rule reads from MongoDB and adds the roles to the token before it is sent to your app.

In that example, the namespace variable would be `http://your_domain.com`
